docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,436 @@
+of the three imported functions. Assuming that we use *weak symbol* to avoid the
+duplication of the wrapper functions (i.e., only one implementation of the functions is picked),
+the invocation of these ECALLs to both enclaves ends up using the same function ids.
+
+See the following host-side code snippet for an example.
+```
+// ECALL to the enclave `foo`.
+common_1_ecall(enclave_foo);
+
+// ECALL to the enclave `bar`.
+common_1_ecall(enclave_bar);
+```
+
+Since the host can have only one implementation of `ecall_common_1_ecall` (the wrapper function),
+both invocations end up using the same function id, say `0`. However, this causes the mismatch
+on the call into the enclave `bar` where the expected id should be `2`.
+
+**Case Study 2: Selectively importing the ECALLs from same EDLs**
+
+Another scenario that both two enclaves selectively import ECALLs from the same EDLs but the
+selections are different.
+
+```
+// foo.edl
+enclave
+{
+    from "common_2.edl" import *;
+
+    trusted
+    {
+        public void foo_ecall();
+    };
+}
+```
+
+```
+// bar.edl
+enclave
+{
+    from "common_2.edl" import common_2_ecall_2;
+
+    trusted
+    {
+        public void bar_ecall();
+    };
+}
+```
+
+The resulting tables will be as follows.
+- `foo` ECALL table
+  ```
+  oe_ecall_func_t __oe_ecalls_table[] = {
+      (oe_ecall_func_t) ecall_common_2_ecall_1,
+      (oe_ecall_func_t) ecall_common_2_ecall_2,
+      (oe_ecall_func_t) ecall_foo_ecall
+  };
+  ```
+- `bar` ECALL table
+  ```
+  oe_ecall_func_t __oe_ecalls_table[] = {
+      (oe_ecall_func_t) ecall_common_2_ecall_2,
+      (oe_ecall_func_t) ecall_bar_ecall
+  };
+  ```
+
+Clearly, we can see the conflict on the function id.
+
+Proposed revision
+-----------
+
+The main goal of this proposal is to address the limitation that different enclaves on the
+same host cannot import the same EDL files (i.e., using the same set of ECALLs).
+To this end, this document proposes to refactor the ECALL calling mechanism as follows.
+
+**Maintain Local and Global ECALL tables on the Host**
+
+- Local Tables
+  Given that the rule of importing `EDL` files is flexible, the resulting
+  ECALL table is not deterministic; i.e., the same ECALL may have different ids
+  across enclaves. To solve the problem, this document uses an idea of storing the
+  list of ECALL id per enclave, which we refer to as local ECALL tables.
+  The local table is declared as follows.
+
+  ```
+  typedef struct _oe_ecall_id_t
+  {
+     uint64_t id;
+  } oe_ecall_id_t;
+
+  typedef struct _oe_enclave
+  {
+     ...
+     /* Per-edl ecall id table. */
+     oe_ecall_id_t ecall_id_table[OE_ECALL_MAX];
+     uint32_t ecall_id_table_size;
+     ...
+  } oe_enclave_t;
+  ```
+
+  The local table allows the host to retrieve the ECALL id per enclave instead
+  of using the hard-coded id in the same ECALL across different enclaves.
+
+- Global Table
+  To guarantee the constant-time look-up, the host needs to use the same id to retrieve
```

Comment:
The local tables above make perfect sense to me.  I still don't follow why a global table or global ids are needed at all though.   You already have constant time lookup with the local table, right?